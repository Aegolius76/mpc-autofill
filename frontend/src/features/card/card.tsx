/**
 * This component is a fundamental UI building block for displaying cards.
 * Displays a card's image, some extra information (its name, its source's name, and its DPI),
 * and has optional props for extending the component to include extra functionality.
 * If being used in a gallery, the previous and next images can be cached for visual smoothness.
 */

import { OnLoadingComplete } from "next/dist/shared/lib/get-img-props";
import Image from "next/image";
import React, {
  memo,
  PropsWithChildren,
  ReactElement,
  useEffect,
  useRef,
  useState,
} from "react";
import BSCard from "react-bootstrap/Card";
import styled from "styled-components";

import { RootState } from "@/app/store";
import { SearchQuery, useAppDispatch, useAppSelector } from "@/common/types";
import { CardDocument } from "@/common/types";
import { Spinner } from "@/components/spinner";
import { setSelectedCardAndShowModal } from "@/features/modals/modalsSlice";
import { selectCardDocumentByIdentifier } from "@/features/search/cardDocumentsSlice";

const HiddenImage = styled(Image)`
  z-index: 0;
  opacity: 0;
`;

const VisibleImage = styled(Image)<{
  imageIsLoading?: boolean;
  showDetailedViewOnClick?: boolean;
  zIndex?: number;
}>`
  z-index: ${(props) => props.zIndex ?? 1};
  &:hover {
    cursor: ${(props) => (props.showDetailedViewOnClick ? "pointer" : "auto")};
  }
  opacity: ${(props) => (props.imageIsLoading ? 0 : 1)};
`;

const OutlinedBSCardSubtitle = styled(BSCard.Subtitle)`
  outline: solid 1px #ffffff00;
  transition: outline 0.2s ease-in-out;
  &:hover {
    outline-color: #ffffffff;
    cursor: pointer;
  }
`;

interface CardImageProps {
  maybeCardDocument: CardDocument | null;
  hidden: boolean;
  small: boolean;
  showDetailedViewOnClick: boolean;
}

function CardImage({
  maybeCardDocument,
  hidden,
  small,
  showDetailedViewOnClick,
}: CardImageProps) {
  //# region queries and hooks

  const dispatch = useAppDispatch();

  //# endregion

  //# region state

  const [imageState, setImageState] = useState<
    "loading" | "loaded" | "errored"
  >("loading");
  const image = useRef<HTMLImageElement>(null);

  //# endregion

  //# region callbacks

  const onLoadingComplete: OnLoadingComplete = (img) => {
    setImageState("loaded");
  };
  const onError: React.ReactEventHandler<HTMLImageElement> = (img) => {
    setImageState("errored");
  };
  const handleShowDetailedView = () => {
    if (showDetailedViewOnClick && maybeCardDocument != null) {
      dispatch(
        setSelectedCardAndShowModal([maybeCardDocument, "cardDetailedView"])
      );
    }
  };

  //# endregion

  //# region effects

  useEffect(() => {
    /**
     * Ensure that the small thumbnail fades in each time the selected image changes.
     * Next.js seems to not fire `onLoadingComplete` when opening a page with a cached image.
     * This implementation was retrieved from https://stackoverflow.com/a/59809184
     */

    setImageState(
      image.current == null || !image.current.complete ? "loading" : "loaded"
    );
  }, [maybeCardDocument?.identifier]);

  //# endregion

  //# region computed constants

  // TODO: always point at image server once it's stable
  const imageCDNURL = process.env.NEXT_PUBLIC_IMAGE_CDN_URL;
  const smallThumbnailURL =
    imageCDNURL != null
      ? `${imageCDNURL}/images/google_drive/small/${maybeCardDocument?.identifier}.jpg`
      : maybeCardDocument?.small_thumbnail_url;
  const mediumThumbnailURL =
    imageCDNURL != null
      ? `${imageCDNURL}/images/google_drive/large/${maybeCardDocument?.identifier}.jpg`
      : maybeCardDocument?.medium_thumbnail_url;
  const imageSrc: string | undefined =
    imageState !== "errored"
      ? small
        ? smallThumbnailURL
        : mediumThumbnailURL
      : small
      ? "/error_404.png"
      : "/error_404_med.png";
  const imageAlt = maybeCardDocument?.name ?? "Unnamed Card";

  //# endregion

  return (
    <>
      {imageState === "loading" && !hidden && <Spinner zIndex={2} />}
      {imageSrc != null &&
        (hidden ? (
          <HiddenImage
            ref={image}
            className="card-img"
            loading="lazy"
            src={imageSrc}
            onLoadingComplete={onLoadingComplete}
            onErrorCapture={onError}
            alt={imageAlt}
            fill={true}
          />
        ) : (
          <VisibleImage
            ref={image}
            className="card-img card-img-fade-in"
            loading="lazy"
            imageIsLoading={imageState === "loading"}
            showDetailedViewOnClick={showDetailedViewOnClick}
            src={imageSrc}
            onLoadingComplete={onLoadingComplete}
            onErrorCapture={onError}
            onClick={handleShowDetailedView}
            alt={imageAlt}
            fill={true}
          />
        ))}
    </>
  );
}

export const MemoizedCardImage = memo(CardImage);

interface CardProportionWrapperProps {
  small: boolean;
  bordered?: boolean;
}

const CardProportionWrapperStyle = styled.div<{ $borderWidth?: number }>`
  z-index: 0;
  background: #4e5d6c;
  border: solid ${(props) => props.$borderWidth ?? 0}px black;
`;

function CardProportionWrapper({
  small,
  bordered = false,
  children,
}: PropsWithChildren<CardProportionWrapperProps>) {
  return (
    <CardProportionWrapperStyle
      $borderWidth={bordered ? 2 : 0}
      className={`rounded-${small ? "lg" : "xl"} shadow-lg ratio ratio-7x5`}
    >
      {children}
    </CardProportionWrapperStyle>
  );
}

export const MemoizedCardProportionWrapper = memo(CardProportionWrapper);

interface CardProps {
  /** The card image identifier to display. */
  maybeCardDocument: CardDocument | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the previous image for visual smoothness. */
  maybePreviousCardDocument?: CardDocument | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the next image for visual smoothness. */
  maybeNextCardDocument?: CardDocument | undefined;
  /** The string to display in the `Card` header. */
  cardHeaderTitle: string;
  /** An element (intended for use with a series of buttons) to include in the `Card` header.  */
  cardHeaderButtons?: ReactElement;
  /** An element (e.g. prev/next buttons) to display in the card footer. If not passed, no footer will be rendered. */
  cardFooter?: ReactElement;
  /** A callback function for when the `Card` (the HTML surrounding the image) is clicked. */
  cardOnClick?: React.MouseEventHandler<HTMLElement>;
  /** A callback function for when the card name is clicked. */
  nameOnClick?: React.MouseEventHandler<HTMLElement>;
  /** The `SearchQuery` specified when searching for this card. */
  searchQuery?: SearchQuery | undefined;
  /** Whether no search results were found when searching for `searchQuery` under the configured search settings. */
  noResultsFound: boolean;
  /** Whether to highlight this card by showing a glowing border around it. */
  highlight?: boolean;
}

export function Card({
  maybeCardDocument,
  maybePreviousCardDocument,
  maybeNextCardDocument,
  cardHeaderTitle,
  cardHeaderButtons,
  cardFooter,
  cardOnClick,
  nameOnClick,
  searchQuery,
  noResultsFound,
  highlight,
}: CardProps) {
  /**
   * This component enables displaying cards with auxiliary information in a flexible, consistent way.
   */

  //# region computed constants

  const cardImageElements =
    maybeCardDocument != null ? (
      <>
        <MemoizedCardImage
          maybeCardDocument={maybeCardDocument}
          hidden={false}
          small={true}
          showDetailedViewOnClick={cardOnClick == null}
        />
        {maybePreviousCardDocument != null &&
          maybePreviousCardDocument.identifier !==
            maybeCardDocument?.identifier && (
            <MemoizedCardImage
              maybeCardDocument={maybePreviousCardDocument}
              hidden={true}
              small={true}
              showDetailedViewOnClick={false}
            />
          )}
        {maybeNextCardDocument != null &&
          maybeNextCardDocument.identifier !==
            maybeCardDocument?.identifier && (
            <MemoizedCardImage
              maybeCardDocument={maybeNextCardDocument}
              hidden={true}
              small={true}
              showDetailedViewOnClick={false}
            />
          )}
      </>
    ) : noResultsFound ? (
      <Image
        className="card-img card-img-fade-in"
        loading="lazy"
        style={{ zIndex: 1 }}
        src="/blank.png"
        alt="Card not found"
        fill={true}
      />
    ) : (
      <Spinner />
    );
  const BSCardSubtitle: typeof BSCard.Subtitle =
    nameOnClick != null ? OutlinedBSCardSubtitle : BSCard.Subtitle;

  //# endregion

  return (
    <BSCard
      className={`mpccard ${highlight ? "mpccard-highlight" : "mpccard-hover"}`}
      onClick={cardOnClick}
      style={{ contentVisibility: "auto" }}
    >
      <BSCard.Header className="pb-0 text-center">
        <p className="mpccard-slot">{cardHeaderTitle}</p>
        {cardHeaderButtons}
      </BSCard.Header>
      <div>
        <MemoizedCardProportionWrapper small={true}>
          {cardImageElements}
        </MemoizedCardProportionWrapper>
        <BSCard.Body className="mb-0 text-center">
          <BSCardSubtitle className="mpccard-name" onClick={nameOnClick}>
            {maybeCardDocument != null && maybeCardDocument.name}
            {maybeCardDocument == null &&
              searchQuery != undefined &&
              searchQuery.query}
          </BSCardSubtitle>
          <div className="mpccard-spacing">
            <BSCard.Text className="mpccard-source">
              {maybeCardDocument != null &&
                `${maybeCardDocument.source_verbose} [${maybeCardDocument.dpi} DPI]`}
              {maybeCardDocument == null &&
                searchQuery != undefined &&
                "Your search query"}
            </BSCard.Text>
          </div>
        </BSCard.Body>
      </div>
      {cardFooter != null && (
        <BSCard.Footer
          className="padding-top"
          style={{ paddingTop: 50 + "px" }}
        >
          {cardFooter}
        </BSCard.Footer>
      )}
    </BSCard>
  );
}

export const MemoizedCard = memo(Card);

interface EditorCardProps {
  /** The card image identifier to display. */
  imageIdentifier: string | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the previous image for visual smoothness. */
  previousImageIdentifier?: string | undefined;
  /** If this `Card` is part of a gallery, use this prop to cache the next image for visual smoothness. */
  nextImageIdentifier?: string | undefined;
  /** The string to display in the `Card` header. */
  cardHeaderTitle: string;
  /** An element (intended for use with a series of buttons) to include in the `Card` header.  */
  cardHeaderButtons?: ReactElement;
  /** An element (e.g. prev/next buttons) to display in the card footer. If not passed, no footer will be rendered. */
  cardFooter?: ReactElement;
  /** A callback function for when the `Card` (the HTML surrounding the image) is clicked. */
  cardOnClick?: React.MouseEventHandler<HTMLElement>;
  /** A callback function for when the card name is clicked. */
  nameOnClick?: React.MouseEventHandler<HTMLElement>;
  /** The `SearchQuery` specified when searching for this card. */
  searchQuery?: SearchQuery | undefined;
  /** Whether no search results were found when searching for `searchQuery` under the configured search settings. */
  noResultsFound: boolean;
  /** Whether to highlight this card by showing a glowing border around it. */
  highlight?: boolean;
}

export function EditorCard({
  imageIdentifier,
  previousImageIdentifier,
  nextImageIdentifier,
  cardHeaderTitle,
  cardHeaderButtons,
  cardFooter,
  cardOnClick,
  nameOnClick,
  searchQuery,
  noResultsFound,
  highlight,
}: EditorCardProps) {
  /**
   * This component is a thin layer on top of `Card` that retrieves `CardDocument` items by their identifiers
   * from the Redux store (used in the project editor).
   * We have this layer because search results are returned as a list of image identifiers
   * (to minimise the quantity of data stored in Elasticsearch), so the full `CardDocument` items must be looked up.
   */

  //# region queries and hooks

  const maybeCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, imageIdentifier)
  );
  const maybePreviousCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, previousImageIdentifier)
  );
  const maybeNextCardDocument = useAppSelector((state: RootState) =>
    selectCardDocumentByIdentifier(state, nextImageIdentifier)
  );

  //# endregion

  return (
    <Card
      maybeCardDocument={maybeCardDocument}
      maybePreviousCardDocument={maybePreviousCardDocument}
      maybeNextCardDocument={maybeNextCardDocument}
      cardHeaderTitle={cardHeaderTitle}
      cardHeaderButtons={cardHeaderButtons}
      cardFooter={cardFooter}
      cardOnClick={cardOnClick}
      nameOnClick={nameOnClick}
      searchQuery={searchQuery}
      noResultsFound={noResultsFound}
      highlight={highlight}
    />
  );
}

export const MemoizedEditorCard = memo(EditorCard);
