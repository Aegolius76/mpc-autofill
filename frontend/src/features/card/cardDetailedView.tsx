import Modal from "react-bootstrap/Modal";
import Row from "react-bootstrap/Row";
import Table from "react-bootstrap/Table";
import { downloadImage, imageSizeToMBString } from "@/common/utils";
import Button from "react-bootstrap/Button";
import React, { useEffect, useState, memo } from "react";
import { useSelector } from "react-redux";
import { RootState } from "@/app/store";
import Image from "next/image";

interface CardDetailedViewProps {
  imageIdentifier: string;
  show: boolean;
  handleClose: {
    (): void;
    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
  };
}

export function CardDetailedView(props: CardDetailedViewProps) {
  // ensure that the medium thumbnail fades in each time the selected image changes
  const [mediumThumbnailLoading, setMediumThumbnailLoading] = useState(true);
  useEffect(() => setMediumThumbnailLoading(true), [props.imageIdentifier]);

  const maybeCardDocument = useSelector(
    (state: RootState) =>
      state.cardDocuments.cardDocuments[props.imageIdentifier]
  );
  const maybeSourceDocuments = useSelector(
    (state: RootState) => state.sourceDocuments.sourceDocuments
  );

  return (
    <div>
      {maybeCardDocument != null && (
        <Modal show={props.show} onHide={props.handleClose} size={"xl"}>
          <Modal.Header closeButton>
            <Modal.Title>Card Details</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            <Row>
              <div
                className="col-lg-5 mb-3 mb-lg-0"
                style={{ position: "relative" }}
              >
                <div
                  className="rounded-xl shadow-lg ratio ratio-7x5"
                  style={{ zIndex: 0 }}
                >
                  <div
                    className="d-flex justify-content-center align-items-center"
                    style={{
                      opacity: mediumThumbnailLoading ? 1 : 0,
                    }}
                  >
                    <div
                      className="spinner-border"
                      style={{ width: 4 + "em", height: 4 + "em" }}
                      role="status"
                    >
                      <span className="visually-hidden">Loading...</span>
                    </div>
                  </div>

                  <Image
                    alt={maybeCardDocument.name}
                    className="card-img-fade-in"
                    style={{
                      zIndex: 1,
                      opacity: mediumThumbnailLoading ? 0 : 1,
                    }}
                    src={maybeCardDocument.medium_thumbnail_url}
                    onLoad={() => setMediumThumbnailLoading(false)}
                    fill={true}
                    // onError={{thumbnail_404(this)}}
                  />
                </div>
              </div>
              <div className="col-lg-7">
                <h4>{maybeCardDocument.name}</h4>
                <Table hover>
                  <tbody>
                    <tr>
                      <td>
                        <b>Source Name</b>
                      </td>
                      <td>
                        {maybeSourceDocuments != null &&
                        maybeSourceDocuments[maybeCardDocument.source_id]
                          .external_link != null ? (
                          <a
                            href={
                              maybeSourceDocuments[maybeCardDocument.source_id]
                                .external_link
                            }
                            target="_blank"
                          >
                            {maybeCardDocument.source}
                          </a>
                        ) : (
                          <a>{maybeCardDocument.source}</a>
                        )}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Source Type</b>
                      </td>
                      <td>{maybeCardDocument.source_type}</td>
                    </tr>
                    <tr>
                      <td>
                        <b>Class</b>
                      </td>
                      <td>
                        {maybeCardDocument.card_type.charAt(0).toUpperCase() +
                          maybeCardDocument.card_type.slice(1).toLowerCase()}
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Identifier</b>
                      </td>
                      <td>
                        <code>{maybeCardDocument.identifier}</code>
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <b>Resolution</b>
                      </td>
                      <td>{maybeCardDocument.dpi} DPI</td>
                    </tr>
                    <tr>
                      <td>
                        <b>Date Created</b>
                      </td>
                      <td>{maybeCardDocument.date}</td>
                    </tr>
                    <tr>
                      <td>
                        <b>File Size</b>
                      </td>
                      <td>{imageSizeToMBString(maybeCardDocument.size, 2)}</td>
                    </tr>
                  </tbody>
                </Table>
                <div className="d-grid gap-0">
                  <Button
                    variant="primary"
                    onClick={() =>
                      downloadImage(maybeCardDocument.download_link)
                    }
                  >
                    Download Image
                  </Button>
                </div>
              </div>
            </Row>
          </Modal.Body>
          <Modal.Footer>
            <Button variant="secondary" onClick={props.handleClose}>
              Close
            </Button>
          </Modal.Footer>
        </Modal>
      )}
    </div>
  );
}

export const MemoizedCardDetailedView = memo(CardDetailedView);
