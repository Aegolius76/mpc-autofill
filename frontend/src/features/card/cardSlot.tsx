/**
 * A higher-level wrapper for the `Card` component with additional functionality.
 * Card slots allow modifying the selected image for the given slot number and face,
 * both via previous/next arrows and the grid selector. Clicking the selected image
 * displays the detailed view. Card slots can be deleted, which also deletes the
 * card slot for the same slot number in the other face.
 */

import React, { useEffect, useState, memo } from "react";
import { useSelector, useDispatch } from "react-redux";
import { AppDispatch, RootState } from "@/app/store";
import { MemoizedCard } from "./card";
import { Back } from "@/common/constants";
import { wrapIndex } from "@/common/utils";
import { deleteImage, setSelectedImage } from "../project/projectSlice";
import Button from "react-bootstrap/Button";
import { MemoizedCardSlotGridSelector } from "./gridSelector";
import { Faces, SearchQuery } from "@/common/types";

interface CardSlotProps {
  searchQuery: SearchQuery | undefined;
  face: Faces;
  slot: number;
  handleShowDetailedView: {
    (selectedImage: string): void;
  };
}

export function CardSlot(props: CardSlotProps) {
  const searchQuery = props.searchQuery;
  const face = props.face;
  const slot = props.slot;

  const [showGridSelector, setShowGridSelector] = useState(false);

  const handleCloseGridSelector = () => setShowGridSelector(false);
  const handleShowGridSelector = () => setShowGridSelector(true);

  const dispatch = useDispatch<AppDispatch>();

  // TODO: move this selector into searchResultsSlice
  // this is a bit confusing. if the card has a query, use the query's results. if it's a cardback with no query,
  // display the common cardback's results.
  const cardbacks =
    useSelector((state: RootState) => state.cardbacks.cardbacks) ?? [];
  const projectCardback = useSelector(
    (state: RootState) => state.project.cardback
  );
  const searchResultsForQueryOrDefault = useSelector((state: RootState) =>
    searchQuery?.query != null
      ? (state.searchResults.searchResults[searchQuery.query] ?? {})[
          searchQuery.card_type
        ]
      : face === Back
      ? cardbacks
      : []
  );

  const projectMember = useSelector(
    (state: RootState) => state.project.members[slot][face]
  );
  const selectedImage = projectMember?.selectedImage;

  const handleShowDetailedView = () => {
    if (selectedImage != null) {
      props.handleShowDetailedView(selectedImage);
    }
  };

  useEffect(() => {
    /**
     * Set the selected image according to some initialisation logic (if search results have loaded).
     */

    if (searchResultsForQueryOrDefault != null) {
      let mutatedSelectedImage = selectedImage;

      // If an image is selected and it's not in the search results, deselect the image
      if (
        mutatedSelectedImage != null &&
        !searchResultsForQueryOrDefault.includes(mutatedSelectedImage)
      ) {
        mutatedSelectedImage = undefined;
      }

      // If no image is selected and there are search results, select the first image in search results
      if (
        searchResultsForQueryOrDefault.length > 0 &&
        mutatedSelectedImage == null
      ) {
        if (searchQuery != null) {
          mutatedSelectedImage = searchResultsForQueryOrDefault[0];
        } else if (face === Back && projectCardback != null) {
          mutatedSelectedImage = projectCardback;
        }
      }

      dispatch(
        setSelectedImage({
          face,
          slot,
          selectedImage: mutatedSelectedImage,
        })
      );
    }
  }, [searchResultsForQueryOrDefault, projectCardback]);

  // const selectedImage = useSelector((state: RootState) => (state.project.members[slot] != null ? (state.project.members[slot][face] != null ? state.project.members[slot][face].selectedImage : null) : null))

  const searchResultsForQuery = searchResultsForQueryOrDefault ?? [];
  const selectedImageIndex: number | undefined =
    selectedImage != null
      ? searchResultsForQuery.indexOf(selectedImage)
      : undefined;
  const previousImage: string | undefined =
    selectedImageIndex != null
      ? searchResultsForQuery[
          wrapIndex(selectedImageIndex + 1, searchResultsForQuery.length)
        ]
      : undefined;
  const nextImage: string | undefined =
    selectedImageIndex != null
      ? searchResultsForQuery[
          wrapIndex(selectedImageIndex - 1, searchResultsForQuery.length)
        ]
      : undefined;

  const deleteThisImage = () => {
    dispatch(deleteImage({ slot }));
  };

  function setSelectedImageFromDelta(delta: number): void {
    // TODO: docstring
    if (selectedImageIndex != null) {
      dispatch(
        setSelectedImage({
          face,
          slot,
          selectedImage:
            searchResultsForQuery[
              wrapIndex(
                selectedImageIndex + delta,
                searchResultsForQuery.length
              )
            ],
        })
      );
    }
  }

  const cardHeaderTitle = `Slot ${slot + 1}`;
  // TODO: the padlock should be replaced by a checkbox showing if the card is selected or not
  const cardHeaderButtons = (
    <>
      <button className="padlock">
        <i className="bi bi-unlock"></i>
      </button>
      <button className="remove">
        <i className="bi bi-x-circle" onClick={deleteThisImage}></i>
      </button>
    </>
  );
  const cardFooter = (
    <>
      {searchResultsForQuery.length === 1 && (
        <p className="mpccard-counter text-center align-middle">
          1 / {searchResultsForQuery.length}
        </p>
      )}
      {searchResultsForQuery.length > 1 && (
        <>
          <Button
            variant="outline-info"
            className="mpccard-counter-btn"
            onClick={handleShowGridSelector}
          >
            {(selectedImageIndex ?? 0) + 1} / {searchResultsForQuery.length}
          </Button>
          <div>
            <Button
              variant="outline-primary"
              className="prev"
              onClick={() => setSelectedImageFromDelta(-1)}
            >
              &#10094;
            </Button>
            <Button
              variant="outline-primary"
              className="next"
              onClick={() => setSelectedImageFromDelta(1)}
            >
              &#10095;
            </Button>
          </div>
        </>
      )}
    </>
  );

  return (
    <div style={{ contentVisibility: "auto" }}>
      <MemoizedCard
        imageIdentifier={selectedImage}
        previousImageIdentifier={previousImage}
        nextImageIdentifier={nextImage}
        cardHeaderTitle={cardHeaderTitle}
        cardFooter={cardFooter}
        cardHeaderButtons={cardHeaderButtons}
        imageOnClick={handleShowDetailedView}
        searchQuery={searchQuery}
        noResultsFound={
          searchResultsForQueryOrDefault != null &&
          searchResultsForQueryOrDefault.length === 0
        }
      />

      {searchResultsForQuery.length > 1 && (
        <MemoizedCardSlotGridSelector
          face={face}
          slot={slot}
          searchResultsForQuery={searchResultsForQuery}
          show={showGridSelector}
          handleClose={handleCloseGridSelector}
        />
      )}
    </div>
  );
}

export const MemoizedCardSlot = memo(CardSlot);
