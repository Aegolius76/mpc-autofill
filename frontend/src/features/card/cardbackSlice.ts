import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { APIGetCardbacks } from "@/app/api";
import { CardbacksState } from "@/common/types";

export const fetchCardbacks = createAsyncThunk(
  "cardbacks/fetchCardbacks",
  async (arg, thunkAPI) => {
    return APIGetCardbacks();
  }
);

const initialState: CardbacksState = {
  cardbacks: [],
  status: "idle",
  error: null,
};

export const cardbackSlice = createSlice({
  name: "cardbacks",
  initialState,
  reducers: {
    addCardbackDocuments: (state, action) => {
      state.cardbacks = [...action.payload];
    },
  },
  extraReducers(builder) {
    builder
      .addCase(fetchCardbacks.pending, (state, action) => {
        state.status = "loading";
      })
      .addCase(fetchCardbacks.fulfilled, (state, action) => {
        state.status = "succeeded";
        state.cardbacks = [...action.payload];
      })
      .addCase(fetchCardbacks.rejected, (state, action) => {
        state.status = "failed"; // TODO: build some stuff for displaying error messages
        state.error = action.error.message ?? null;
      });
  },
});

export default cardbackSlice.reducer;
