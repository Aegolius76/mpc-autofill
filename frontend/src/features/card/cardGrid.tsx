/**
 * This component displays all `CardSlot`s in the project and is responsible for
 * querying the server for search results as necessary.
 */

import React, { useEffect, useMemo, useState } from "react";
import Modal from "react-bootstrap/Modal";
import Row from "react-bootstrap/Row";
import styled from "styled-components";

import { Back, Front } from "@/common/constants";
import { useAppDispatch, useAppSelector } from "@/common/types";
import { selectBackendURL } from "@/features/backend/backendSlice";
import { MemoizedCardSlot } from "@/features/card/cardSlot";
import { selectProjectMembers } from "@/features/project/projectSlice";
import { fetchCardDocumentsAndReportError } from "@/features/search/cardDocumentsSlice";
import { clearSearchResults } from "@/features/search/searchResultsSlice";
import { Spinner } from "@/features/ui/spinner";

const CardGridDefaultBackground = styled.div`
  position: absolute;
  border: 1px darkgray solid;
  border-radius: 10px;
  height: 100%;
  width: 100%;
  box-shadow: 0 0 50px 0 black inset;
`;

const CardGridDefaultText = styled.div`
  position: relative;
  font-size: 1.25em;
  text-align: center;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-shadow: 0 4px 15px #000000;
  max-width: 90%;
`;

function CardGridDefault() {
  return (
    <CardGridDefaultBackground>
      <CardGridDefaultText>
        <p>Your project is empty at the moment.</p>
        <p>
          Use the <b>Add Cards</b> menu on the right to get started!
        </p>
      </CardGridDefaultText>
    </CardGridDefaultBackground>
  );
}

export function CardGrid() {
  const dispatch = useAppDispatch();

  const fetchingCardData = useAppSelector(
    (state) =>
      state.cardDocuments.status == "loading" ||
      state.searchResults.status === "loading"
  );

  const backendURL = useAppSelector(selectBackendURL);

  const cardSlotsFronts = [];
  const cardSlotsBacks = [];
  const projectMembers = useAppSelector(selectProjectMembers);
  const searchSettings = useAppSelector((state) => state.searchSettings);
  const stringifiedSearchSettings = useMemo<string>(
    () => JSON.stringify(searchSettings),
    [searchSettings]
  );

  useEffect(() => {
    // recalculate search results when search settings change
    if (backendURL != null && searchSettings.sourceSettings.sources != null) {
      dispatch(clearSearchResults());
      fetchCardDocumentsAndReportError(dispatch);
    }
  }, [stringifiedSearchSettings]);

  for (const [slot, slotProjectMember] of projectMembers.entries()) {
    cardSlotsFronts.push(
      <MemoizedCardSlot
        key={`${Front}-slot${slot}`}
        searchQuery={slotProjectMember.front?.query}
        face={Front}
        slot={slot}
      ></MemoizedCardSlot>
    );
    cardSlotsBacks.push(
      <MemoizedCardSlot
        key={`${Back}-slot${slot}`}
        searchQuery={slotProjectMember.back?.query}
        face={Back}
        slot={slot}
      ></MemoizedCardSlot>
    );
  }

  const frontsVisible = useAppSelector(
    (state) => state.viewSettings.frontsVisible
  );

  // TODO: should we aim to lift state up here and conditionally render rather than hide?
  return (
    <>
      {projectMembers.length == 0 && <CardGridDefault />}
      <Row
        xxl={4}
        lg={3}
        md={2}
        sm={1}
        xs={1}
        className="g-0"
        style={{ display: frontsVisible ? "" : "none" }}
      >
        {cardSlotsFronts}
      </Row>
      <Row
        xxl={4}
        lg={3}
        md={2}
        sm={1}
        xs={1}
        className="g-0"
        style={{ display: frontsVisible ? "none" : "" }}
      >
        {cardSlotsBacks}
      </Row>
      <Modal show={fetchingCardData} centered>
        <Modal.Header>
          <Modal.Title
            style={{ textAlign: "center", userSelect: "none" }}
            className="w-100"
          >
            Loading your cards...
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Spinner />
        </Modal.Body>
      </Modal>
    </>
  );
}
