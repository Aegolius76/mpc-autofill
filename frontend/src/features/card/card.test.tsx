import { renderWithProviders } from "@/common/test-utils";
import { Card } from "@/features/card/card";
import { screen, render } from "@testing-library/react";

test("the html structure of a Card with a single identifier", () => {
  const rendered = renderWithProviders(
    <Card
      cardHeaderTitle="Card 1"
      noResultsFound={false}
      imageIdentifier="abc123"
    />,
    {
      preloadedState: {
        cardDocuments: {
          cardDocuments: {
            abc123: {
              identifier: "abc123",
              card_type: "CARD",
              name: "Card Name",
              priority: 0,
              source: "Card Source",
              source_id: 0,
              source_verbose: "Card Source",
              source_type: "google drive",
              dpi: 1200,
              searchq: "card name",
              extension: "png",
              date: "1st January, 2000", // formatted by backend
              download_link: "",
              size: 10_000_000,
              small_thumbnail_url: "",
              medium_thumbnail_url: "",
            },
          },
          status: "idle",
          error: null,
        },
      },
    }
  );
  expect(rendered.baseElement).toMatchSnapshot();
});
