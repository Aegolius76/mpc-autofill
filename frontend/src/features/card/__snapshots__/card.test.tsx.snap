// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of a Card with a single identifier 1`] = `
<body>
  <div>
    <div
      class="mpccard mpccard-hover card"
    >
      <div
        class="pb-0 text-center card-header"
      >
        <p
          class="mpccard-slot"
        >
          Card 1
        </p>
      </div>
      <div>
        <div
          class="rounded-lg shadow-lg ratio ratio-7x5"
          style="z-index: 0;"
        >
          <div
            class="d-flex justify-content-center align-items-center"
            style="opacity: 1;"
          >
            <div
              class="spinner-border"
              role="status"
              style="width: 4em; height: 4em;"
            >
              <span
                class="visually-hidden"
              >
                Loading...
              </span>
            </div>
          </div>
          <img
            alt="Card Name"
            class="card-img card-img-fade-in"
            loading="lazy"
            src=""
            style="z-index: 1; opacity: 0;"
          />
        </div>
        <div
          class="mb-0 text-center card-body"
        >
          <div
            class="mpccard-name card-subtitle h6"
          >
            Card Name
          </div>
          <div
            class="mpccard-spacing"
          >
            <p
              class="mpccard-source card-text"
            >
              Card Source [1200 DPI]
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
